File: app/src/main/java/com/monk/utils/BitmapUtils.java
package com.monk.utils;

import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;

/**
 * @author monk
 * @date 2019-01-21
 */
public class BitmapUtils {

    public Bitmap decodeSampleBitmapFromResource(Resources res, int resId, int reqWidth, int reqHeight) {
        BitmapFactory.Options options=new BitmapFactory.Options();
        options.inJustDecodeBounds=true;
        BitmapFactory.decodeResource(res, resId, options);
        options.inSampleSize = calculateInSampleSize(options, reqWidth, reqHeight);
        options.inJustDecodeBounds=false;
        return BitmapFactory.decodeResource(res,resId,options);

    }

    private int calculateInSampleSize(BitmapFactory.Options options, int reqWidth, int reqHeight) {
        int inSampleSize=1;
        int rawWidth = options.outWidth;
        int rawHeight = options.outHeight;

        if (rawHeight > reqHeight || rawWidth > reqWidth) {
            int halfRawHeight = rawHeight / 2;
            int halfRawWidth = rawWidth / 2;
            while (halfRawHeight / inSampleSize >= reqHeight && halfRawWidth / inSampleSize >= reqWidth) {
                inSampleSize *= 2;
            }
        }
        return inSampleSize;
    }
}

File: commonutils/src/main/java/com/monk/commonutils/BitmapUtils.kt
package com.monk.commonutils

import android.content.res.Resources
import android.graphics.Bitmap
import android.graphics.BitmapFactory

/**
 * @author monk
 * @date 2019-01-21
 */
class BitmapUtils {
    fun decodeSampleBitmapFromResource(res: Resources?, resId: Int, reqWidth: Int, reqHeight: Int): Bitmap {
        val options = BitmapFactory.Options()
        options.inJustDecodeBounds = true
        BitmapFactory.decodeResource(res, resId, options)
        options.inSampleSize = calculateInSampleSize(options, reqWidth, reqHeight)
        options.inJustDecodeBounds = false
        return BitmapFactory.decodeResource(res, resId, options)
    }

    private fun calculateInSampleSize(options: BitmapFactory.Options, reqWidth: Int, reqHeight: Int): Int {
        var inSampleSize = 1
        val rawWidth = options.outWidth
        val rawHeight = options.outHeight
        if (rawHeight > reqHeight || rawWidth > reqWidth) {
            val halfRawHeight = rawHeight / 2
            val halfRawWidth = rawWidth / 2
            while (halfRawHeight / inSampleSize >= reqHeight && halfRawWidth / inSampleSize >= reqWidth) {
                inSampleSize *= 2
            }
        }
        return inSampleSize
    }
}